### PR TITLE
complete CreateKey_napi 、Encrypt_napi、Decrypt_napi、GenerateDataKey_napi

### PR DESCRIPTION
**Pull request by Liang1XZhao**
currently it can support the below REST API cmds:
POST <ehsm_srv_address>/ehsm?Action=CreateKey
POST <ehsm_srv_address>/ehsm?Action=Encrypt
POST <ehsm_srv_address>/ehsm?Action=Decrypt
POST <ehsm_srv_address>/ehsm?Action=GenerateDatakey

test: passed the unit-test
run the script: python test_aes.py

### CODE REVIEW

File: ehsm_kms_service/ehsm_kms_params.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      type: 'string',
+      maxLength: _6kb_length,
+      minLength: 1
+    },
+    aad,
+  },
+  GenerateDataKey: {
+    cmk_base64: {
+      type: 'string',
+      maxLength: _6kb_length,
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_params.js:67]: the datakey's max length is 1024

File: core/App/ehsm_napi.cpp [hunk 2]
```diff
@@ -52,50 +58,53 @@ static char* StringToChar(std::string str)
...
-    void addData(std::string key, std::string data)
-    {
+	
+	void setCode(int code){code = code;};
+	void setMessage(string message){msg = message;};
+	void addData(string key, uint32_t data) {
+        if(jsonStr.size() > 0){
+            jsonStr += ",";
+        }
+        jsonStr += "\""+key+"\" : " + "\""+intToString(data)+"\"";
```
Comment (yang8621) [inline core/App/ehsm_napi.cpp:76]: refer to line@87 std::to_string(code)

File: core/App/ehsm_napi.cpp [hunk 3]
```diff
@@ -31,15 +31,21 @@
...
 #include "serialize.h"
 
-//#include "ehsm_provider.h"
-
 using namespace std;
 using namespace EHsmProvider;
 
-static char* StringToChar(std::string str)
+
+string intToString(int value)
```
Comment (yang8621) [inline core/App/ehsm_napi.cpp:40]: remove it.

File: ehsm_kms_service/ehsm_kms_server.js [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+/**
+ * router
+ */
+app.post('/ehsm', function (req, res) {
+  const PAYLOAD = req.body.payload;
+  
+  // ACTION: request function name
+  const ACTION = req.query.Action;
+  const keyspec = ehsm_keyspec_t[PAYLOAD.keyspec];
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_server.js:246]: why not put this param (const keyspec = ehsm_keyspec_t[PAYLOAD.keyspec];) under 
if(ACTION === apis.CreateKey) {
...
}
?

File: ehsm_kms_service/ehsm_kms_server.js [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+  // todo sign 
+  next();
+}
+/**
+ * check payload
+ */
+ const _checkPayload = function (req,res,next){
+   const action = req.query.Action;
+  const { payload } = req.body;
+  const currentPayLoad = napiparams[action];
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_server.js:55]: name not change?
const currentPayLoad = napiparams[action];

File: ehsm_kms_service/ehsm_kms_params.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+const ehsm_keyorigin_t = {
+  EH_INTERNAL_KEY: 0,
+  EXTERNAL_KEY: 1,
+}
+const keyspec= {
+  type : 'const',
+  arr: Object.keys(ehsm_keyspec_t),
+  errortext: "The keyspec type is incorrect",
+};
+
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_params.js:21]: keep the define align with "origin"
const keyspec= {
  type : 'const',
  arr: Object.keys(ehsm_keyspec_t),
  errortext: "The keyspec type is incorrect",
};

File: ehsm_kms_service/test/test_aes.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+# todo: remove hardcoded parameters
+# test CreateKey(AES)
+data = {
+    "appid":"t123",
+    "nonce":"t123",
+    "timestamp":"t123",
+    "sign":"t123",
+    "payload":{
+        "keyspec":"EHM_AES_GCM_128",
```
Comment (syan10) [inline ehsm_kms_service/test/test_aes.py:15]: not change 
"keyspec":"EHM_AES_GCM_128",

File: core/App/main.cpp [hunk 8]
```diff
@@ -671,8 +649,6 @@ int main(int argc, char* argv[])
 
     testExportDataKey();
 
-    test_NAPI_AES();
```
Comment (syan10) [inline core/App/main.cpp:674]: so there's no backend c++ unittest for the NAPI now, right?

**Comment by yang8621** [comment]
need to add "apt install libjson-c-dev" anywhere?

**Comment by yang8621** [comment]
> need to add "apt install libjson-c-dev" anywhere?

OK. Seems it not used in this commit.

File: core/App/main.cpp [hunk 9]
```diff
@@ -35,14 +35,27 @@
...
+        int len = str.size();
+        retChar = (char *)malloc(len * sizeof(uint8_t));
+        if(retChar != nullptr){
+            memset(retChar, 0, len);
+            memcpy(retChar, str.c_str(), len);
+        }
+    }
+    return retChar;
+}
+
```
Comment (syan10) [inline core/App/main.cpp:58]: remove it

File: ehsm_kms_service/ehsm_kms_server.js [hunk 10]
```diff
@@ -0,0 +1,278 @@
+const express = require('express')
+const ffi = require('ffi-napi');
+const app = express();
+const logger = require('./logs/logger');
+const {params: napiparams, ehsm_keyspec_t, ehsm_keyorigin_t} = require('./ehsm_kms_params.js')
```
Comment (syan10) [inline ehsm_kms_service/ehsm_kms_server.js:5]: change the name

**Pull request closed without merging**